zerver/tests/test_auth_backends.py

```diff
@@ -1505,6 +1507,148 @@ def test_github_oauth2_success_single_email(self) -> None:
+        self.assertEqual(data['subdomain'], 'zulip')
+        self.assertEqual(data['next'], '/user_uploads/image')
+        self.assertEqual(result.status_code, 302)
+        parsed_url = urllib.parse.urlparse(result.url)
+        uri = "{}://{}{}".format(parsed_url.scheme, parsed_url.netloc,
+                                 parsed_url.path)
+        self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
+
+    def test_github_oauth2_login_multiple_accounts_exist(self) -> None:
+        # In the login flow, if multiple of the user's verified emails
+        # are associated with existing accounts, we expect the choose
+        # email screen to select which account to use.
+        account_data_dict = dict(email='hamlet@zulip.com', name="Hamlet")
+        email_data = [
+            dict(email=account_data_dict["email"],
+                 verified=True),
+            dict(email='hamlet@zulip.com',
+                 verified=True,
+                 primary=True),
+            dict(email="aaron@zulip.com",
+                 verified=True),
+            dict(email="ignored@example.com",
+                 verified=False),
+        ]
+        result = self.social_auth_test(account_data_dict,
+                                       subdomain='zulip', email_data=email_data,
+                                       expect_choose_email_screen=True,
+                                       next='/user_uploads/image')
+        data = load_subdomain_token(result)
+        self.assertEqual(data['email'], 'hamlet@zulip.com')
+        self.assertEqual(data['name'], 'Hamlet')
+        self.assertEqual(data['subdomain'], 'zulip')
+        self.assertEqual(data['next'], '/user_uploads/image')
+        self.assertEqual(result.status_code, 302)
+        parsed_url = urllib.parse.urlparse(result.url)
+        uri = "{}://{}{}".format(parsed_url.scheme, parsed_url.netloc,
+                                 parsed_url.path)
+        self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))
+
+    def test_github_oauth2_login_no_account_exists(self) -> None:
+        # In the login flow, if the user has multiple verified emails,
+        # none of which are associated with an existing account, the
+        # choose email screen should be shown (which will lead to a
+        # "continue to registration" choice).
+        account_data_dict = dict(email="not-hamlet@zulip.com", name="Not Hamlet")
+        email_data = [
+            dict(email=account_data_dict["email"],
+                 verified=True),
+            dict(email="notprimary@zulip.com",
+                 verified=True),
+            dict(email="verifiedemail@zulip.com",
+                 verified=True),
+        ]
+        result = self.social_auth_test(account_data_dict,
+                                       subdomain='zulip',
+                                       email_data=email_data,
+                                       is_signup='0',
```

Comment:
Wouldn't this be better with `` is_signup='1'``?

Reply:
yes, I thought it would go to the signup page (we only need to check login flow) if its 1 but it's the same in every case other than `None` for `is_signup`. The test is failing(in `stage_two_of_registration`) if is_signup is not provided, what could be the reason?

Reply:
In `stage_two_of_registration` at `result = self.client_get(result.url)` we are getting `result.status_code = 200`(should be 302) indicating that the user is logged In, I'm unable to figure out why this is happening. might there be some bug in `social_auth_test`?

Reply:
We go to a different URL if you do ``social_auth_test`` without the is_signup arg, see ``prepare_login_url_and_headers`` function:
```
        if is_signup is not None:
            url = self.SIGNUP_URL
```

So the flow will go a bit different and probably that's why it makes the test fail.

Reply:
but the `result` in the following part of code should take us to the registration flow instead of logging in for an email that's not associated with an existing account.
```
            if expect_choose_email_screen and result.status_code == 200:
                # For authentication backends such as GitHub that
                # successfully authenticate multiple email addresses,
                # we'll have an additional screen where the user selects
                # which email address to login using (this screen is a
                # "partial" state of the python-social-auth pipeline).
                #
                # TODO: Generalize this testing code for use with other
                # authentication backends; for now, we just assert that
                # it's definitely the GitHub authentication backend.
                self.assert_in_success_response(["Select account"], result)
                assert self.AUTH_FINISH_URL == "/complete/github/"

                result = self.client_get(self.AUTH_FINISH_URL,
                                         dict(state=csrf_state, email=account_data_dict['email']), **headers)
```
I could not find a way to fix this, could you please suggest a fix for this?

Reply:
Have you verified that the user is indeed getting logged in? (because i agree, I think that shouldn't be happening) Or perhaps it's some other response being given from the server?

Reply:
Earlier the test was verifying if the user was getting logged In instead of taking to registration flow which Tim pointed out. So, I'm thinking it's getting logged in.

Reply:
Could you post the previous code of this test that was verifying logging in and passing?

Reply:
```
    def test_github_oauth2_login_no_account_exists(self) -> None:
         # In the login flow, if the user has multiple verified emails,
         # none of which are associated with an existing account, the
         # choose email screen should be shown (which will lead to a
         # "continue to registration" choice).
         account_data_dict = dict(email='not-hamlet@zulip.com', name="Not Hamlet")
         email_data = [
             dict(email=account_data_dict["email"],
                  verified=True),
             dict(email="notprimary@zulip.com",
                  verified=True),
             dict(email="verifiedemail@zulip.com",
                  verified=True),
         ]
         result = self.social_auth_test(account_data_dict,
                                        subdomain='zulip',
                                        email_data=email_data,
                                        expect_choose_email_screen=True,
                                        next='/user_uploads/image')
         data = load_subdomain_token(result)
         self.assertEqual(data['email'], account_data_dict["email"])
         self.assertEqual(data['name'], account_data_dict["name"])
         self.assertEqual(data['subdomain'], 'zulip')
         self.assertEqual(data['next'], '/user_uploads/image')
         self.assertEqual(result.status_code, 302)
         parsed_url = urllib.parse.urlparse(result.url)
         uri = "{}://{}{}".format(parsed_url.scheme, parsed_url.netloc,
                                  parsed_url.path)
         self.assertTrue(uri.startswith('http://zulip.testserver/accounts/login/subdomain/'))

Reply:
Ah! No, this actually should be fine - this doesn't imply the user will get immediately logged in. If the account needs to be registered we go through the /subdomain/ endpoint, and that will take the user to the registration form. (But yes, the new form of the test is better, because it's actually testing the registration flow).

Reply:
The flow is basically:
``social pipeline -> redirect_and_log_into_subdomain -> log_into_subdomain -> login_or_register_remote_user -> login the user or go to the registration flow depending on what's needed``.

Reply:
So do I need to change anything now other than rebasing?

Reply:
Then what's happening when  is_signup=None, I think it's not registering as `stage_two_of_registration` is failing.
if it's getting logging in, there is probably some big that needs to be fixed.

Reply:
> Then what's happening when is_signup=None

The flow goes through through /login/ instead /register/, which means steps are a bit different (I believe in going through ``confirm_continue_registration``), so the test isn't in the flow where ``stage_two_of_registration`` expects when it's called.

Reply:
What I noticed however that the ``prepare_login_url_and_headers`` function doesn't care about the value of ``is_signup`` arg, it goes through the register link either way, as long as it's not ``None``, which is confusing. Perhaps it's worth fixing that in a new commit, perhaps changing the ``is_signup``argument  to boolean in the natural way, and changing all occurences of ``is_signup='1'`` to ``is_signup=True`` in the social tests. (``is_signup='0'`` isn't used anywhere fortunately based on my quick check)